### ISSUE TITLE
BUG: autoreload 

### ISSUE THREAD

**Issue by jbwhit** [ipython/ipython#10395]
@dmargala and I think that we have found a bug in the expected behavior from the [autoreload](http://ipython.readthedocs.io/en/stable/config/extensions/autoreload.html) extension. 

They are related and can be seen (and reproduced) in the following short Jupyter Notebooks: 

[Bug 1](https://github.com/dmargala/ccds-import-demo/blob/master/notebooks/autoreload%20bug%201.ipynb) -- Expected "Reload all modules imported with %aimport every time before executing the Python code typed." --> did not happen.

[Bug 2](https://github.com/dmargala/ccds-import-demo/blob/master/notebooks/autoreload%20bug%202.ipynb) -- Expected "Reload all modules (except those excluded by %aimport) every time before executing the Python code typed." --> also did not happen.

If you want to clone the demo repo it's [here](https://github.com/dmargala/ccds-import-demo).

**Comment by Carreau** [comment]
Hi !

Thanks for your report. I'm digging into it. 

Bug 1. 

This is actually not a bug. You `%aimport` only `src`, you need to `%aimport src.neato` to also mark it as having to be reloaded. I understand your POV, and actually it is a strange feature of Python. 

The fact that when you `import src` the value `src.neato` is set is a side-effect of the Python import machinery. Technically you should alway also import submodules if you are going to use them explicitly. I don't know a lot of modules like that, but there is an (in)famous one. 

```
$ ipython
import Python 3.6.0 | packaged by conda-forge | (default, Feb 10 2017, 07:08:35)
Type 'copyright', 'credits' or 'license' for more information
IPython 6.0.0.dev -- An enhanced Interactive Python. Type '?' for help.

In [1]: import matplotlib

In [2]: matplotlib.finance
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
<ipython-input-2-f3835d0d6a64> in <module>()
----> 1 matplotlib.finance

AttributeError: module 'matplotlib' has no attribute 'finance'

In [3]: import matplotlib.finance

In [4]: matplotlib.finance
Out[4]: <module 'matplotlib.finance' from '/Users
```

Which technically is one of the only one doing the supposedly right thing. 

One of the reasoning being that if two submodules depends on each other (even indirectly) then you have to make a placeholder value for some amount of time, but anyway...

Also let say I add the following in your `__init__.py`:

```
from .neato import *
from matplotlib import pyplot # added this.
```

Then `src.pyplot.things` is reasonable. Do you though expect `matplotlib` to be reloaded ? Well even if you might, `%aimport` have to draw the boundary and when autoreload is set to 1 we use the `Module` definition that Python give us. It it tells us that `src` and `src.neato` are different. They just happened to have nested names.

I think I understand the behavior you expect, I'm unsure we can get it to work reliably. Maybe using modules `__name__`, but I can see a few way this could fails. So this will likely be a nothing to fix. 

Will look more  into the second bug later, but it seem you are not excluding anything from reload. 
Should it be `%aimport -src.neato` ?  (note the `-` ).

**Comment by jbwhit** [comment]
Ok, I think I got it. I wrote up a summary of my understand and some examples here: 

https://gist.github.com/jbwhit/38c1035c48cdb1714fc8d47fa163bfae

**Comment by Carreau** [comment]
> Ok, I think I got it. I wrote up a summary of my understand and some examples here:
> 
> https://gist.github.com/jbwhit/38c1035c48cdb1714fc8d47fa163bfae

Yes, that seem more correct. I agree that the documentation is not perfect, and that code has likely not been in use in a long time. 

Do you want to attempt sending a patch that clarifies it ?

**Comment by jbwhit** [comment]
Sure, that sounds good. I'll make an attempt this week. First, should I report somewhere else that the link on the [autoreload docs page](http://ipython.readthedocs.io/en/stable/config/extensions/autoreload.html) saying [Edit on GitHub](https://github.com/ipython/ipython/blob/feb880d80ee6e1d72f11cbef35a907ebf2d390c6/docs/source/config/extensions/autoreload.rst) is broken?

**Comment by Carreau** [comment]
That's a know issue of readthedocs, because we auto generate the `.rst` files from the source files. SO the test is actually [directly in `autoreload.py`](https://github.com/ipython/ipython/blob/master/IPython/extensions/autoreload.py), that allow to get the docs interactively as well.
